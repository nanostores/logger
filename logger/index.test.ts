import { delay } from 'nanodelay'
import {
  atom,
  cleanStores,
  deepMap,
  map,
  STORE_UNMOUNT_DELAY
} from 'nanostores'
import { afterEach, beforeAll, expect, it, vi } from 'vitest'

import { action } from '../index.js'
import { format } from '../test/index.js'
import { logger } from './index.js'

let $atom = atom()
let $map = map({
  artworks: 213,
  fullname: 'Nikolay Suetin',
  id: 'A10',
  username: 'suetin'
})
let $deepMap = deepMap<{
  artists: {
    [key: string]: {
      artworks: string[]
      movement: null | string
    }
  }
}>({
  artists: {
    malevich: {
      artworks: [],
      movement: null
    }
  }
})

let increaseCounter = action($atom, 'Increase Counter', store => {
  store.set(Number(store.get()) + 1)
})

let changeUserArtworks = action(
  $map,
  'Change User Artworks',
  (store, value) => {
    store.setKey('artworks', value)
  }
)

let changeUser = action(
  $map,
  'Change User',
  async (store, username = 'chashnik', fullname = 'Ilya Chashnik') => {
    await delay(1000)
    store.setKey('username', username)
    store.setKey('fullname', fullname)
  }
)

let brokenAction = action($map, 'Broken Throw', async () => {
  throw Error('Something went wrong in the action Throw Error')
})

let addArtworks = action(
  $deepMap,
  'Add Artworks',
  async (store, artist: string, artworks: string[]) => {
    for (let item of artworks) {
      // await delay(100)
      let index = store.get().artists.malevich.artworks.length
      store.setKey(`artists.${artist}.artworks[${index}]`, item)
    }
    throw Error('Something went wrong in action Add Artworks')
  }
)

let out = ''
let groups = 0

beforeAll(() => {
  vi.spyOn(console, 'groupCollapsed').mockImplementation((...args: any[]) => {
    console.log(...args)
    groups += 1
  })
  vi.spyOn(console, 'log').mockImplementation((...args: any[]) => {
    if (out === '') out += '\n'
    out += format(groups, [], ...args) + '\n'
  })
  vi.spyOn(console, 'groupEnd').mockImplementation(() => {
    groups -= 1
  })
})

afterEach(() => {
  out = ''
  groups = 0
  cleanStores($atom, $map, $deepMap)
})

it('prints logs', async () => {
  let destroy = logger({ $atom, $deepMap, $map })

  let unbindAtom = $atom.listen(() => {})
  $atom.set(100)
  $atom.set(101)
  increaseCounter()
  increaseCounter()
  increaseCounter()
  unbindAtom()
  await delay(STORE_UNMOUNT_DELAY + 10)

  changeUserArtworks(303)

  await changeUser()
  await changeUser('malevich', 'Kazimir Malevich')

  try {
    await brokenAction()
  } catch {}

  $deepMap.setKey('artists.malevich.movement', 'Suprematism')

  try {
    await addArtworks('malevich', [
      'White on White',
      'Suprematist Composition',
      'Black Circle'
    ])
  } catch {}
  await delay(STORE_UNMOUNT_DELAY + 10)

  expect(out).toMatchSnapshot()

  destroy()
})

it('returns unbind function', () => {
  let destroy = logger({ $atom })

  destroy()
  $atom.set(100)

  expect(out).toBe('')
})

it('handles and throws error in action', async () => {
  let destroy = logger({ $map })

  await expect(async () => await brokenAction()).rejects.toThrowError()
  expect(out).toMatchSnapshot()

  destroy()
})

it('has option to disable mount logs', async () => {
  let destroy = logger({ $atom }, { messages: { mount: false } })

  $atom.listen(() => {})()
  await delay(STORE_UNMOUNT_DELAY + 10)
  expect(out).toMatchSnapshot()

  destroy()
})

it('has option to disable unmount logs', async () => {
  let destroy = logger({ $atom }, { messages: { unmount: false } })

  $atom.listen(() => {})()
  await delay(STORE_UNMOUNT_DELAY + 10)
  expect(out).toMatchSnapshot()

  destroy()
})

it('has option to disable change logs', async () => {
  let destroy = logger({ $atom, $map }, { messages: { change: false } })

  let unbindAtom = $atom.listen(() => {})
  let unbindMap = $map.listen(() => {})
  $atom.set(100)
  $map.setKey('artworks', 302)
  changeUserArtworks(303)
  unbindAtom()
  unbindMap()
  await delay(STORE_UNMOUNT_DELAY + 10)
  expect(out).toMatchSnapshot()

  destroy()
})

it('has option to disable action logs', async () => {
  let destroy = logger({ $atom, $map }, { messages: { change: false } })

  let unbindAtom = $atom.listen(() => {})
  let unbindMap = $map.listen(() => {})
  $atom.set(100)
  $map.setKey('artworks', 302)
  changeUserArtworks(303)
  unbindAtom()
  unbindMap()
  await delay(STORE_UNMOUNT_DELAY + 10)
  expect(out).toMatchSnapshot()

  destroy()
})

it('has option to disable actions with specific name', async () => {
  let destroy = logger({ $atom, $map }, { ignoreActions: ['Change User'] })

  let unbindMap = $map.listen(() => {})
  await changeUser('puni', 'Ivan Puni')
  changeUserArtworks(56)
  unbindMap()
  await delay(STORE_UNMOUNT_DELAY + 10)
  expect(out).toMatchSnapshot()

  destroy()
})
