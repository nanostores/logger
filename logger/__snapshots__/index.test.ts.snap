// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`has option to disable change logs 1`] = `
"
Nano Stores mount: $atom store was mounted
Nano Stores mount: $map store was mounted
Nano Stores unmount: $atom store was unmounted
Nano Stores unmount: $map store was unmounted
"
`;

exports[`has option to disable mount logs 1`] = `
"
Nano Stores unmount: $atom store was unmounted
"
`;

exports[`has option to disable unmount logs 1`] = `
"
Nano Stores mount: $atom store was mounted
"
`;

exports[`prints logs 1`] = `
"
Nano Stores mount: $atom store was mounted
Nano Stores change: $atom store was changed
  new: 100
Nano Stores change: $atom store was changed
  new: 101
  old: 100
Nano Stores change: $atom store was changed
  new: 102
  old: 101
Nano Stores change: $atom store was changed
  new: 103
  old: 102
Nano Stores change: $atom store was changed
  new: 104
  old: 103
Nano Stores unmount: $atom store was unmounted
Nano Stores change: $map store was changed in the artworks key
  value: 213 → 303 
  new: {"artworks":303,"fullname":"Nikolay Suetin","id":"A10","username":"suetin"}
  old: {"artworks":213,"fullname":"Nikolay Suetin","id":"A10","username":"suetin"}
Nano Stores change: $map store was changed in the username key
  value: suetin → chashnik 
  new: {"artworks":303,"fullname":"Nikolay Suetin","id":"A10","username":"chashnik"}
  old: {"artworks":303,"fullname":"Nikolay Suetin","id":"A10","username":"suetin"}
Nano Stores change: $map store was changed in the fullname key
  value: Nikolay Suetin → Ilya Chashnik 
  new: {"artworks":303,"fullname":"Ilya Chashnik","id":"A10","username":"chashnik"}
  old: {"artworks":303,"fullname":"Nikolay Suetin","id":"A10","username":"chashnik"}
Nano Stores change: $map store was changed in the username key
  value: chashnik → malevich 
  new: {"artworks":303,"fullname":"Ilya Chashnik","id":"A10","username":"malevich"}
  old: {"artworks":303,"fullname":"Ilya Chashnik","id":"A10","username":"chashnik"}
Nano Stores change: $map store was changed in the fullname key
  value: Ilya Chashnik → Kazimir Malevich 
  new: {"artworks":303,"fullname":"Kazimir Malevich","id":"A10","username":"malevich"}
  old: {"artworks":303,"fullname":"Ilya Chashnik","id":"A10","username":"malevich"}
Nano Stores change: $deepMap store was changed in the artists.malevich.movement key
  new: {"artists":{"malevich":{"artworks":["Black Square"],"movement":"Suprematism"}}}
  old: {"artists":{"malevich":{"artworks":["Black Square"],"movement":null}}}
Nano Stores mount: $deepMap store was mounted
Nano Stores change: $deepMap store was changed in the artists.malevich.artworks[1] key
  new: {"artists":{"malevich":{"artworks":["Black Square","White on White"],"movement":"Suprematism"}}}
  old: {"artists":{"malevich":{"artworks":["Black Square"],"movement":"Suprematism"}}}
Nano Stores change: $deepMap store was changed in the artists.malevich.artworks[2] key
  new: {"artists":{"malevich":{"artworks":["Black Square","White on White","Suprematist Composition"],"movement":"Suprematism"}}}
  old: {"artists":{"malevich":{"artworks":["Black Square","White on White"],"movement":"Suprematism"}}}
Nano Stores change: $deepMap store was changed in the artists.malevich.artworks[3] key
  new: {"artists":{"malevich":{"artworks":["Black Square","White on White","Suprematist Composition","Black Circle"],"movement":"Suprematism"}}}
  old: {"artists":{"malevich":{"artworks":["Black Square","White on White","Suprematist Composition"],"movement":"Suprematism"}}}
Nano Stores unmount: $deepMap store was unmounted
"
`;
